Macos network extension port

majestrate opened the pull request:
add the required bullshit to make macos work without root.

TODO:
- [x] Need abstraction around system-specific route poking, out of the VPNPlatform interface.  (Will be coming as a separate PR).  This is required for:
  - [x] add/remove default routes when exit is turned on/off
  - [x] add support for selective IP range exit mode routing
- [x] fix DNS leak when exit mode is enabled (The issue: the appex doesn't appear to be respecting the configured routes through its own tun device, so DNS lookups don't go through the exit).

Commit:
initial network extension code for macos
probably does not work

Commit:
add swift version bullshit file and additional bullshittery

jagerman committed:
Various cmake build cleanups/refactors

- Added contrib/macos/README.txt with description of the cancer
  happening here.
- Add provisioningprofiles that Apple wants to make things work properly
- Made the entitlements files match the provisioningprofiles
- Remove configured entitlements files; we *can't* change any of the
  things here because they are closedly tied to the provisioningprofiles
  -- which means if someone wants to build their own Lokinet, they have
  to replace a bunch of crap and change application IDs throughout.
  This is the hostile-to-open-source Apple way.
- Remove unused old lokinet binary, as we're no longer using it on macos
- Use a POST_BUILD rather than install to copy things around into the
  right places
- Convert all the configure_file's to consistently use @ONLY
- Misc cleanups

jagerman committed:
Set version into Info.plist, don't manually configure

cmake already treats the info plist as a file to be configured (not
merely copied) so we don't need to configure_file ourselves to a temp
file.

jagerman committed:
Switch extension from a framework to an appex

jagerman committed:
Default CODESIGN_APPEX to CODESIGN_APP

Sometimes (e.g. dev builds) these can apparently be the same; sometimes
it seems they need to be different, because Apple.

Commit:
macos sort of works now

Commit:
make ci run ./contrib/mac.sh with no codesigning

Commit:
add dummy sign target for ci when we don't have signing keys
remove static macos from ci pipeline

Commit:
it works

jagerman committed:
Consolidate apple-specific bits into llarp/apple

jagerman committed:
Tweak icns generation to only run when necessary

jagerman committed:
Apple OS interface cleanup & refactoring

- Add a C callback interface (context_wrapper.h) between lokinet and the
  objective-C code so that:
  - we can use objective-C (rather than objective-C++), which seems more
    likely to be supported by Apple into the future;
  - we minimize the amount of code that needs to be aware of the Apple
    APIs.
  - this replaces apple logger objective c++ implementation with a plain
    c++ implementation that takes a very simple C callback (provided
    from the obj-c code) to actually make the call to NSLog.

- Add various documentation to the code of what is going on.

- Send all DNS traffic to the primary IP on the tun interface.  The
  match prefixes simply don't work as advertised, and have weird shit
  (like even if you get it working for some domains, "instagram.com"
  still doesn't because of god-knows-what Apple internal politics).

- Drop the dns proxy code as we don't need it anymore.

- Don't use 9.9.9.9 for default DNS.  (We might consider the unfiltered
  9.9.9.10 as an alternative default, but if we do it should be a global
  lokinet change rather than a Mac-specific change).

- Parse a lokinet.ini in the data directory, if it exists.  (Since we
  are sandboxed, it is an app-specific "home" directory so is probably
  buried god knows where, but at least the GUI ought to be able to get
  it to let users add things to it).

- This commit also adds a swift version of the PacketTunnelProvider
  glue, which ought to work in theory, but the *tooling* for cmake is so
  underdeveloped that I couldn't find any way to actually get the damn
  thing working.  So I'm committing it here anyway (and will revert it
  away in the next commit) in case we someday want to switch to it.

-

jagerman committed:
Remove swift version (do not squash)

Don't squash this commit so that the swift version stays around in
history in case we need to resurrect it again some day (i.e. when Apple
decides to kill off Objective-C support).

jagerman committed:
format.sh: support macports clang-format

macports names it clang-format-mp-N to avoid clashing with a
system-installed one.

jagerman commented:
Various updates:

- DNS works now, without requiring exit mode.

- Add a C callback interface (context_wrapper.h) between lokinet and the objective-C code so that:
  - we can use objective-C (rather than objective-C++), which seems more likely to be supported by Apple into the future;
  - we minimize the amount of code that needs to be aware of the Apple APIs.
  - this replaces apple logger objective c++ implementation with a plain c++ implementation that takes a very simple C callback (provided from the obj-c code) to actually make the call to NSLog.

- Add various documentation to the code of what is going on.

- Send all DNS traffic to the primary IP on the tun interface.  The match prefixes simply don't work as advertised, and have weird shit (like even if you get it working for some domains, "instagram.com" still doesn't because of god-knows-what Apple internal politics).

- Drop the dns proxy code as we don't need it anymore.

- Don't use 9.9.9.9 for default DNS.  (We might consider the unfiltered 9.9.9.10 as an alternative default, but if we do it should be a global lokinet change rather than a Mac-specific change).

- Parse a lokinet.ini in the data directory, if it exists.  (Since we are sandboxed, it is an app-specific "home" directory so is probably buried god knows where, but at least the GUI ought to be able to get it to let users add things to it).

- Attempted a swift version, which should have worked fine in theory, but the tooling around Swift outside of XCode is too immature to be useful -- basically Swift is so underdeveloped outside Apple that you're pretty much stuck properly something non-trivial -- such as a C bridging class in an App extension -- outside of XCode, and XCode is its own pile of warts that would break our ability to do CI. So I converted the swift version back to Objective C which is much more mature and works fine.  (The swift version is added in one commit and removed in the next, to preserve it in case we need it in the future.)

jagerman committed:
Remove no-longer-used framework.mm

jagerman committed:
Generate a default client lokinet.ini on startup if it doesn't exist

Thus when a user goes looking for it they'll find the (commented out)
default in the right place and can edit it.

(That right place is: ~/Library/Containers/com.loki-project.lokinet.network-extension/Data/lokinet.ini)

jagerman committed:
Remove more unused dns-proxy files

jagerman committed:
Remove unused/empty/no longer used apple headers

jagerman committed:
Add lokinet subnet to default routes

Mac doesn't route to a tunnel's ip range by default.  WTF.

jagerman committed:
Add RouteManager; make exit on/off work

jagerman committed:
Remove unused vars

jagerman committed:
Add DNS trampoline

This runs a DNS listener on localhost:1053 that bounces requests to the
upstream DNS through the tunnel.  The idea here is that, when we turn on
exit mode, we start libunbound bouncing the requests through the
trampoline (since if it makes direct requests they won't go through the
tunnel).

(The actual libunbound configuration is still to follow).

jagerman committed:
Remove debug

jagerman committed:
Unleak exit mode DNS via unbound DNS trampoline on (macOS)

When we enable/disable exit mode on this restarts the unbound DNS
responder with the DNS trampoline (or restores upstream, when disabling)
to properly route DNS requests through the tunnel (because libunbound's
direct requests don't get tunneled because unbound is inside the network
extension).

jagerman committed:
Reformat

jagerman committed:
Silence warnings

jagerman committed:
Remove dead code

majestrate committed:
disable language ASM, yolo

jagerman committed:
Forward-declare ub_ctx/ub_result

Avoids needing unbound.h in the search path to include the
unbound_resolver.hpp header.

majestrate committed:
invert logic for android dns hook

jagerman commented on CMakeLists.txt:
```diff
@@ -29,10 +41,10 @@ if(RELEASE_MOTTO AND CMAKE_BUILD_TYPE MATCHES "[Rr][Ee][Ll][Ee][Aa][Ss][Ee]")
   add_definitions(-DLLARP_RELEASE_MOTTO="${RELEASE_MOTTO}")
 endif()
 
+set(LOKINET_VERSION "${lokinet_VERSION_MAJOR}.${lokinet_VERSION_MINOR}.${lokinet_VERSION_PATCH}")
```
Don't need to set a variable for this: ${lokinet_VERSION} will already be set to this (by the project(...) call).

jagerman commented on cmake/macos_installer_deps.cmake:
```diff
@@ -1,8 +1,9 @@
 # macos specific cpack stuff goes here
 
+return()
```
This looks temporary: I think we should just delete this file (and follow it back to remove includes).

jagerman commented on contrib/lokinet.svg:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Generator: Adobe Illustrator 25.0.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
```
Remove stupid Adobe comment

jagerman commented on contrib/mac.sh:
```diff
@@ -0,0 +1,39 @@
+      -DWITH_SYSTEMD=OFF \
+      -DFORCE_OXENMQ_SUBMODULE=ON \
+      -DSUBMODULE_CHECK=OFF \
+      -DWITH_LTO=OFF \
```
If LTO works we should remove this to let LTO be used (I'm testing it now).

jagerman commented on daemon/CMakeLists.txt:
```diff
@@ -1,7 +1,16 @@
+if(APPLE)
+  set(LOKINET_SWIFT_SOURCES lokinet.swift)
+  add_executable(lokinet ${LOKINET_SWIFT_SOURCES})
+  configure_file(${CMAKE_CURRENT_SOURCE_DIR}/Lokinet.modulemap.in ${CMAKE_CURRENT_BINARY_DIR}/swift/LokinetExtension/module.modulemap ESCAPE_QUOTES @ONLY)
```
I think we may not need this modulemap file anymore (TIAS).

jagerman commented on daemon/Lokinet.modulemap.in:
```diff
@@ -0,0 +1,4 @@
+
+module LokinetExtension {
+       header "@CMAKE_SOURCE_DIR@/include/lokinet-extension.hpp"
+}
```
This header doesn't exist anymore so I'm pretty sure we can remove this modulemap.

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -11,6 +11,10 @@
+#ifdef __APPLE__
+#include <llarp/util/logging/apple_logger.hpp>
+#endif
+
```
We don't build this at all on apple anymore, should drop these changes (this and the one below).

jagerman commented on include/lokinet-dnsproxy.hpp:
```diff
@@ -0,0 +1,19 @@
+
+- (BOOL)handleNewFlow:(NEAppProxyFlow*)flow;
+
+@end
```
Obsolete, delete.

jagerman commented on include/lokinet-extension.hpp:
```diff
@@ -0,0 +1,21 @@
+- (void)handleAppMessage:(NSData*)messageData
+       completionHandler:(void (^)(NSData* responseData))completionHandler;
+
+@end
```
Obsolete, delete.

jagerman commented on llarp/config/config.cpp:
```diff
@@ -1446,4 +1446,17 @@ namespace llarp
+    return config;
+  }
+#endif
+
```
Obsolete, can be dropped (the apple code now uses/parses a full lokinet.ini config).

jagerman commented on llarp/config/config.hpp:
```diff
@@ -260,6 +260,11 @@ namespace llarp
+    static std::shared_ptr<Config>
+    NetworkExtensionConfig(std::string exit);
+#endif
+
```
Obsolete, can be dropped.

jagerman commented on llarp/dns/server.cpp:
```diff
@@ -120,12 +127,11 @@ namespace llarp::dns
-    return true;
+#ifdef ANDROID
+    return IsUpstreamResolver(to, from);
+#else
```
This looks suspicious to me:

When android (configured with random DNS server) makes a DNS request, we want to intercept it, hand it over to unbound, then unbound makes the request to the lokinet-configured upstream.

So we want this to return true on the first request (to intercept) and false on the second request (to let the upstream request through).

And that is what the `#else` case is going to do; the logical inversion for android doesn't seem right.

(Either way, this needs a comment.  I suggest:
```C++
// If this request is going to an upstream resolver then we want to let it through (i.e. don't handle it), and so
// want to return false.  If we have something else then we want to intercept it to route it through our caching
// libunbound server (which then redirects the request to the lokinet-configured upstream, if not cached).
```

jagerman commented on llarp/handlers/tun.cpp:
```diff
@@ -61,6 +61,21 @@ namespace llarp
+      IsUpstreamResolver(const SockAddr& to, const SockAddr& from) const override
+      {
+        LogError(
+            "IsUpstreamResolver? ", to.asIPv6(), " != ", m_Endpoint->GetIfAddr(), ", from=", from);
```
Debug statement should get deleted.

jagerman committed:
Re-enable LTO; target macos 10.12+

jagerman committed:
Remove debugging

jagerman committed:
LOKINET_VERSION -> built in lokinet_VERSION

jagerman committed:
Remove old mac installer packaging

jagerman committed:
Remove advertising of big crappy company

jagerman committed:
Remove obsolete bigs

Removes stuff we didn't end up needing/using:
- Lokinet.modulemap
- apple bits from lokinet.cpp (we don't use lokinet.cpp at all on macos
  anymore).
- dnsproxy/extension C++ headers
- apple-specific network extension config in llarp::config::Config

majestrate committed:
add comment about android dns

majestrate committed:
tmp commit for debugging (revert me)

majestrate committed:
remove unused import

lokinet.swift no longer touches internals of lokinet network extension so we remove the import for it

jagerman committed:
Fix IPv4 routes are disabling exit

(This was actually already working because macOS is buggy and doesn't
actually remove routes other than the default).

majestrate committed:
override ShouldHandlePacket on android to bypass non .snode/.loki dns hooking.

jagerman committed:
Add IPv6 addr/routing for macos

majestrate committed:
set ipv6 address on apple to hardcoded value

majestrate committed:
format
